File: viper.c
#include "viper.h"

int main(int argc, char *argv[]) {
   char *file = 0;			// filename passwordfile
   char pass[MAXENCPWDLENGTH+1] = "";	// encrypted password
   char *user = 0;			// username in passwordfile
   char *lsf  = 0;			// filename loadsourcefile
   char *pf   = 0;			// filename progressfile
   int  rf    = 0;			// runtime limit
   int  chr   = 0;			// characterset
   int  pws   = 1;			// min passwordlength
   int  pwl   = DEFAULTPWLENGTH;	// max passwordlength
   int  ui    = 10;			// console update interval
   int  vo    = 0;			// verbose output
   int  i     = 0;			// loop variable
   FILE *fp_lsf;			// loadsourcefile
   FILE *fp_file;			// passwordfile
   FILE *fp_cset;			// character set file
   char line[255];			// tmp buffer
   char vp_stat[255];			// last saved status
   char * linebuf = NULL;

   printf("\nViper v1.6 (Hale 05/12/2000) - C version by Frank4DD (05/05/2014)\n");
   printf("Wiltered Fire - www.wilter.com/wf, incl. bugfixes by David C. Rankin\n\n");
	
   /* need help? */
   if ( argc == 1 || (!(strcmp (argv[1], "-h"))) || (!(strcmp (argv[1], "-?"))) ) {
      help();
   }
	
   /* verbose output on? */
   for (i = 1; i < argc; i++) {
      if (! (strcmp (argv[i], "-v"  ))) { vo = 1; i++;}
   }
	
   if (vo) {
      if ( (argc != 2) && (argc != 4) && (argc != 6) &&
           (argc != 8) && (argc != 10) && (argc != 12) 
            && (argc != 14) && (argc != 16) ) {
         printf("missing value for argument: try viper -h\n");
         exit(-1);
      }
   }
   else {
      if ( (argc != 1) && (argc != 3) && (argc != 5) &&
           (argc != 7) && (argc != 9) && (argc != 11) 
	    && (argc != 13) && (argc != 15) ) {
         printf("missing value for argument: try viper -h\n");
         exit(-1);
      }
   }
	
   /* process command line arguments */
   for (i = 1; i < argc; i++) { 	
      if (! (strcmp (argv[i], "-f"  ))) { file =      argv[i+1] ; i++;}
      else if (! (strcmp (argv[i], "-u"  ))) { user =      argv[i+1] ; i++;}
      else if (! (strcmp (argv[i], "-c"  ))) { chr  = atoi(argv[i+1]); i++;}
      else if (! (strcmp (argv[i], "-pwl"))) { pwl  = atoi(argv[i+1]); i++;}
      else if (! (strcmp (argv[i], "-ui" ))) { ui   = atoi(argv[i+1]); i++;}
      else if (! (strcmp (argv[i], "-pws"))) { pws  = atoi(argv[i+1]); i++;}
      else if (! (strcmp (argv[i], "-lsf"))) { lsf  =      argv[i+1] ; i++;}
      else if (! (strcmp (argv[i], "-pf" ))) { pf   =      argv[i+1] ; i++;}
      else if (! (strcmp (argv[i], "-rf" ))) { rf   = atoi(argv[i+1]); i++;}
      else if (! (strcmp (argv[i], "-v"  ))) { continue; }
      else { printf("Unknown argument \"%s\": try viper -h\n", argv[i]); exit(-1); }
   }
	
   /* break early if calling from file */
   if (lsf) {
      if ( (fp_lsf = fopen(lsf, "r+")) == NULL ) {
         printf("Error: Can't open %s!\n", lsf);
         exit(-1);
      }
      fscanf (fp_lsf, "%s", vp_stat);
		
      /* check to see if run has been completed */
	
      if (! (strcmp (vp_stat, FIN_IDENT))) {
         printf("The saved run has been completed.\n");
         printf("Check %s for details.\n", lsf);
         fclose(fp_lsf);
         exit(-1);
      }
			
      /* continue otherwise */
      fscanf (fp_lsf, "%d", &lsf_out.ci_pws);
      fscanf (fp_lsf, "%d", &lsf_out.ci_pwl);
      fscanf (fp_lsf, "%s", lsf_out.ci_pass);
      fscanf (fp_lsf, "%s", lsf_out.ci_user);
      fscanf (fp_lsf, "%s", lsf_out.ci_dnum);
      fscanf (fp_lsf, "%s", lsf_out.ci_cset);
      fscanf (fp_lsf, "%s", lsf_out.ci_pf);
      fscanf (fp_lsf, "%d", &lsf_out.ci_ui);
      fclose(fp_lsf);
      lsf_out.ci_rf = 0;
      lsf_out.ci_vo = vo;
      printf("...loaded parameters from file %s.\n", lsf);
      crack(lsf_out);
   }
	
   /* check for required arguments */
   if (!file) {
      printf("Error: Password filename required!\n");
      exit(-1);
   }
   else if (!user) {
      printf("Error: Username required!\n");
      exit(-1);
   }
	
   /* attempt to load account from file */
   else if ( (fp_file = fopen(file, "r")) == NULL ) {
      printf("Error: Can't open %s!\n", file);
      exit(-1);
   }
		
   while ( (fscanf (fp_file, "%s", line) != EOF) ) {
      if (! (strcmp (user, strtok(line, ":"))) ) {
         linebuf = strtok(NULL, ":");
         if ( !linebuf ) {
            printf("No password for user %s!\n", user);
            exit(-1);
         }
         strcpy(pass, linebuf);
         if ( !pass || (strlen(pass)) <  4 ) {
            printf("Error: Bad password for user %s!\n", user);
            exit(-1);
         }
         printf("Found: user %s pw: %s\n", user, pass);
      }
   }
   fclose(fp_file);
   if ( !pass || (strlen(pass)) <  4 ) {
      printf("Error: No user %s in file %s!\n", user, file);
      exit(-1);
   }

   /* load character set */
   if ( (fp_cset = fopen(CHARSET_FILE, "r")) == NULL ) {
      printf("Error: Can't open %s!\n", CHARSET_FILE);
      exit(-1);
   }

   while ( (fscanf (fp_cset, "%s", line) != EOF) ) {
      if ( chr == (atoi(line)) ) {
         fscanf (fp_cset, "%s", lsf_out.ci_cset);
          break;
      }
   }
	
   if ( !lsf_out.ci_cset || (strlen(lsf_out.ci_cset)) < 2 ) {
      printf("Error: Bad charset %d in %s!\n", chr, CHARSET_FILE);
      exit(-1);
   }
   else {
      printf("Found: Charset %d in %s\n", chr, CHARSET_FILE);
   }
	
   fclose(fp_cset);
	
   /* write data in struct */
   lsf_out.ci_rf = rf;
   if (pf) { strcpy (lsf_out.ci_pf, pf); }
   lsf_out.ci_pws = pws;
   lsf_out.ci_pwl = pwl;
   strcpy (lsf_out.ci_pass, pass);
   strcpy (lsf_out.ci_user, user);
   lsf_out.ci_ui = ui;
   lsf_out.ci_vo = vo;
   printf("...command line parameters loaded.\n");
   crack(lsf_out);
}

/* ######## start subs ############### */

/* program help() function */
int help() {
   printf("\t-f <file>    File to load password from (required unless using lsf)\n");
   printf("\t-u <user>    Username to load from file (required unless using lsf)\n");
   printf("\t-lsf <file>  Load saved file from previous session\n");
   printf("\t-pf <file>   Save progress to file at update interval\n");
   printf("\t-rf #        Amount of time in hours to run for (default infinite)\n");
   printf("\t-c #         Character set from charset.ini to use (default 1)\n");
   printf("\t-pws #       Minimum password length (starting value, default 1)\n");
   printf("\t-pwl #       Maximum password length (default %d - maximum %d)\n", 
   	  DEFAULTPWLENGTH, MAXPASSWDLENGTH);
   printf("\t-ui #        Console update interval (in minutes - default 10)\n");
   printf("\t-v           Verbose output\n");
   exit(0);
}

/* The results function the_res() */
void the_res(struct crack_input erg, char * endpass, struct tm start) {
   static struct tm act_time;
   double end_time;
   int r;
   char convert_str[17];
	
   time(&read_time);
   act_time=*localtime(&read_time);
   end_time = difftime(mktime(&act_time), mktime(&start));
   convert(end_time, convert_str);
	
   sprintf(message[0], "%s\n", FIN_IDENT);
   sprintf(message[1], "\n");
	
   if (endpass[0] != '\0') {
      sprintf(message[2], " The password has been located.\n");
      sprintf(message[3], " Username : %s\n", erg.ci_user);
      sprintf(message[4], " Password : %s\n", endpass);
   }
   else {
      sprintf(message[2], " The password could not be located.\n");
      sprintf(message[3], " Username : %s\n", erg.ci_user);
      sprintf(message[4], " Password : %s\n", "** unknown **");
   }
	
   sprintf(message[5], " Started  : %s", asctime(&start));
   sprintf(message[6], " Finished : %s", asctime(&act_time));
   sprintf(message[7], " Duration : %s\n", time_done);
	
   for (r = 1; r <= 7; r++) { printf("%s", message[r]); }

   if(strlen(lsf_out.ci_pf)) {
      if ( (fp_pf = fopen(lsf_out.ci_pf, "w")) == NULL ) {
         printf("Error: Can't open %s!\n", lsf_out.ci_pf);
         exit(-1);
      }
      for (r = 0; r <=7; r++) { fprintf(fp_pf, message[r]); }
      fclose(fp_pf);
   }
	
   printf("\nViper exiting...\n");
   exit(0);
}

File: viper.pl
#!/usr/local/bin/perl

print "\nViper v1.4 \(Hale 5/12/2000\)\nWiltered Fire - www.wilter.com/wf\n\n";

$arg_length = @ARGV;
$file = "";
$user = "";
$char = 1;
$pws = 1;
$pwl = 8;      
$ui = 10;
$pf = "";

# need help?

if($ARGV[0] eq "-h" || $ARGV[0] eq "-?" || $ARGV[0] eq "") { &help; }

# process command line arguments

for($i=0;$i<$arg_length;$i++) {
 if($ARGV[$i] eq "-f") { $file = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-u") { $user = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-c") { $char = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-pwl") { $pwl = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-ui") { $ui = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-pws") { $pws = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-lsf") { $lsf = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-pf") { $pf = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-rf") { $runfor = $ARGV[$i+1]; $i++; }
 if($ARGV[$i] eq "-v") { $verbose = 1; }
 }


# handle minimum value (v1.3)
$pws--;

# break early if calling from file

if($lsf) {
 open(LSF,$lsf) || die "Could not open saved file $lsf\!\n";
 $vp_stat = <LSF>; chop($vp_stat);
 
 ## check to see if run has been completed # 
 if($vp_stat eq "--viper_final--") { die "The saved run has been completed - check $lsf for details\n\n"; }

 #continue otherwise

 $pws   = <LSF>; chop($pws);
 $pwl   = <LSF>; chop($pwl);
 $epass = <LSF>; chop($epass);
 $user  = <LSF>; chop($user);
 $nump  = <LSF>; chop($nump);
 $vars  = <LSF>; chop($vars);
 $pf    = <LSF>; chop($pf);
 $ui    = <LSF>; chop($ui);
 @vars = split(/ /,$vars);
 $charset = join("",@vars);
 close(LSF);
 
 print " \[Loaded parameters from file $lsf\]\n";

 &crack($user,$epass,$charset,$runfor,$pwl,$pws,$ui,1,$nump,$pf);
 }

# check for required arguments

if($file eq "" || $user eq "") {
 print "Both the password file and user name are required\n";
 die "\n";
 }

# attempt to load account from file

open(IN,$file) || die "Could not open password file $file\n";
$epass = "";
while(<IN>) {
 chop($_);
 @sort = split(/:/,$_);
 if(@sort[0] eq $user) { $epass = @sort[1]; }
 }
close(IN);
if($epass eq "") { die "User $user could not be found in $file\n"; }
$pl = length($epass);
if($pl < 4) { die "Bad password for user $user\n"; } # eliminate shadowed 
if($verbose eq 1) { print "Found user $user \(pw:$epass\)\n"; }

# load character set

open(IN,"charset.ini") || die "Could not open charset.ini\n";
$charset = "";
while(<IN>) {
 chop($_);
 @sort = split(/ /,$_);
 if(@sort[0] eq $char) { $charset = @sort[1]; }
 }
close(IN);
if($charset eq "") { die "Could not load character set index $char\n"; }

&crack($user,$epass,$charset,$runfor,$pwl,$pws,$ui,0,"",$pf);

######## start subs ###############

## begin help sub ##

sub help {
 print "   -f <file>    File to load password from (required unless using lsf)\n";
 print "   -u <user>    Username to load from file (required unless using lsf)\n\n";
 print "   -lsf <file>  Load saved file from previous session\n";
 print "   -pf <file>   Save progress to file at update interval\n";
 print "   -rf #        Amount of time in minutes to run for (default forever)\n\n";
 print "   -c #         Character set from charset.ini to use (default 1)\n";         
 print "   -pws #       Minimum password length (starting value - default 1)\n";
 print "   -pwl #       Maximum password length (default 8)\n";
 print "   -ui #        Console update interval (in minutes - default 10)\n";
 print "   -v           Verbose output\n";
 die "\n";
 }

## begin crack sub ##

sub crack {
 my($user,$epass,$charset,$runfor,$pwl,$pws,$ui,$lsf,$nump,$pf) = @_;

 @vars = split(//,$charset);
 $varlen = @vars;
 $count = 0;
 $refresh = 0;
 $realpass = "";
 $on = 0;
 $ui = $ui*60;
 $runfor = $runfor*60;

 # confirm runfor
 if($runfor) {
 $trunfor = &convert($runfor);
 @tsort = split(/:/,$trunfor); # only want hh:mm
 print " \[Running for @tsort[0]\:@tsort[1]\]\n"; }

 # get current time and get initial time since 1/1/1970
 $sloctime = localtime();
 $stime = time();
 $ltime = $stime;
 $cur_time = $stime;

 # display info and start 
 if($verbose eq 1) { print "\nCharacter set $char loaded with $varlen character length\n"; }
 print "Starting crack on $sloctime\n\n";

 # go on with the show....
 
 if($lsf eq 1) { @pass = split(/\,/,$nump); $on = 1;}

 $salt = substr($epass,0,2); 

 for($i=$pws;$i<$pwl;$i++) {
  ### make new password ###
  $it = $i+1;
  
  # initalize 500k counter
  $fivek = 0;
  
  # initalize the password array #
  if($on eq 1) { $on = 0; } else { for($j=0;$j<$it;$j++) { @pass[$j] = 0; } }
  
  # calculate number of password for this length  
  $tot_num = $mult = $varlen;
  for($tc=1;$tc<$it;$tc++) { $tot_num *= $mult; }

  print "Cracking for pass length $it";
  if($verbose eq 1) { print " \($tot_num possibilities\)"; }
  print "\n";

  ### start loop for this length ###
  while(@pass[0] ne $varlen) {
   $count++;
   $refresh++;   

   ### make actual password out of array ###
   $pass = "";
   for($y=0;$y<@pass;$y++) { $pass = $pass . @vars[@pass[$y]]; }

   ## - done - ##
   
   ### Here is where the magic happens ###
   $cpass = crypt($pass,$salt); # the point of the whole script
   if($epass eq $cpass) { 
    $realpass = $pass;
    the_res($user,$realpass,$stime,$sloctime,$pf); 
    }         

   ## - time check - ##
   if($count eq 500000) { 
    $rtime = time(); 
    $time_dif = $rtime - $cur_time; 
    $uicount = $uicount + $count;
    $count=0;
    $fivek++;
    }
   
   ## - ui check - ##
   if($time_dif >= $ui) {
    # calculate duration and cps
    $ctime = time();
    $cur_time = $ctime;
    $totdur = $ctime - $stime;
    $temp_dur = $ctime - $ltime;
    $ltime = $ctime;
    $cps = int $uicount / $temp_dur;

    # v1.4 add-on / time-remaining and percent calculation #
    $pre = int $tot_num / 500000;
    $percent = $fivek / $pre;
    $percent = int $percent * 10000;
    $percent /= 100;
    
    $left = $tot_num - ($fivek*500000);
    $i_sec = int $left / $cps;
    $etaf = &convert($i_sec);
   
    #change second format into hh:mm:ss

    $dur = &convert($totdur);

    #display data

    print "\n[ Length: $it \| Percent: $percent\% \| CPS: $cps \| Time : $dur ]\n";
    if($verbose eq 1) { print "[ Estimated Time Remaining : $etaf \| Last: $pass]\n"; }
    
    ## additional conditional routine : auto-save progress (v1.3)

    if($pf ne "") {
     $numpass = join("\,",@pass);
     $uia = $ui / 60;
     open(TMP,">$pf") || print STDOUT "Could not open progress file $pf!\n";
     print TMP "--viper_prog--\n$i\n$pwl\n$epass\n$user\n$numpass\n@vars\n$pf\n$uia\n";
     close(TMP);
     }
    
    ## check for -rf expiration
    if($runfor) {
    if($totdur >= $runfor) { die "\n \[ RunFor duration expired \]\n"; } }
    
    ## update time counter
    $uicount = 0;
    $time_dif = 0;
    }
   ## whee... ##   

   ### create next password array ###
   $l = @pass;
   $l--;
   @pass[$l]++;   # wow.. this is an important line... 
   for($r=$l;$r>0;$r--) { if(@pass[$r] eq $varlen) { @pass[$r-1]++; @pass[$r] = 0; } }
   ## done with that ###
   }
  }
 the_res($user,$realpass,$stime,$sloctime,$pf);
 }

## begin results sub ##

sub the_res {
my($user,$realpass,$stime,$sloctime,$pf) = @_;
 
if($realpass ne "") { 
  $loctime = localtime();
  $etime = time();
  $ttime = $etime - $stime;
  $duration = &convert($ttime);
  $message = "
  The password has been located!
  Username : $user 
  Password : $realpass
  Started  : $sloctime
  Finished : $loctime
  Time     : $duration
  ";
  } else {
  $message = "
  The password could not be located
  Username : $user 
  Password : ???
  Started  : $sloctime
  Finished : $loctime
  Time     : $duration
  ";
   }
  print "$message\n";

if($pf ne "") {
 open(FINAL,">$pf") ;
 print FINAL "--viper_final--\n\n$message";
 close(FINAL);
 }

die "Viper exiting...\n";
}

## begin time convert sub (understands days/years v1.4) ##

sub convert {
 my($csec) = @_;
 $cmin = 0; $chour = 0; $cday = 0; $cyear = 0;

 while($csec >= 31536000) { $cyear++; $csec -= 31536000; }
 while($csec >= 86400) { $cday++; $csec -= 86400; }
 while($csec >= 1440) { $chour++; $csec -= 1440; }
 while($csec >= 60) { $cmin++; $csec -= 60; }
 
 $a = length($csec);
 if($a eq 1) { $csec = "0" . $csec; }
 if($a eq 0) { $csec = "00"; }
 $a = length($cmin);
 if($a eq 1) { $cmin = "0" . $cmin; }
 if($a eq 0) { $cmin = "00"; }
 $a = length($chour);
 if($a eq 1) { $chour = "0" . $chour; }
 if($a eq 0) { $chour = "00"; }

 $dur = "";
 if($cyear ne 0) { $dur = $cyear . "y "; }
 if($cday ne 0) { $dur = $dur . $cday . "d "; }
 $dur = $dur . $chour . ":" . $cmin . ":" . $csec;
 
 return $dur;
 }
